core/trino-spi/src/main/java/io/trino/spi/type/DecimalAggregationAccumulatorType.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.spi.type;
+
+import io.airlift.slice.Slice;
+import io.trino.spi.block.Block;
+import io.trino.spi.block.BlockBuilder;
+import io.trino.spi.block.BlockBuilderStatus;
+import io.trino.spi.block.DecimalAggregationAccumulatorBlockBuilder;
+import io.trino.spi.block.PageBuilderStatus;
+import io.trino.spi.connector.ConnectorSession;
+
+import static java.util.Collections.singletonList;
+
+public class DecimalAggregationAccumulatorType
```

Comment:
We don't want to add synthetic types for this. Why can't this be represented as a ROW type?

Reply:
I just saw the earlier discussion. I'd like to understand why using ROW types is slower.

Reply:
So I tried RowType and it looked very slow for this case (some details below).
Then I tried generic FixedBinaryType (like Varbinary but with fixed entry length). It is fast for serialization and deserialization but it always sends overflow data over the network even if all overflow values are 0.
That's how this dedicated type was born, It's more or less as fast for SerDe as FixedBinary and optimizes the network for the overflow case.

**So Why RowType is slow?**
It looks like RowType forces the serializer to do more work that is not necessary e.g. resizing multiple fields separately instead of two arrays.
I created row type serializer in a separate branch https://github.com/trinodb/trino/pull/10311 + added it to the benchmark.
Below are the results.
If you want to dig in deeper here are flame graphs for the runs: [deserialize](https://lukasz-stec.github.io/trino/jmh/2/io.trino.operator.aggregation.state.BenchmarkLongDecimalWithOverflowAndLongStateSerializer.deserialize-AverageTime-nullRate-0-overflowRate-0-serializerClass-LongDecimalWithOverflowAndLongStateSerializerRowType/flame-cpu-forward.html) and [serialize](https://lukasz-stec.github.io/trino/jmh/2/io.trino.operator.aggregation.state.BenchmarkLongDecimalWithOverflowAndLongStateSerializer.serialize-AverageTime-nullRate-0-overflowRate-0-serializerClass-LongDecimalWithOverflowAndLongStateSerializerRowType/flame-cpu-forward.html)

I suspect we could try to optimize rowType for this use case but it's gonna be hard to beat both dedicated block and even varbinary. 
```
Benchmark                                                                  (nullRate)  (overflowRate)                                     (serializerClass)  Mode  Cnt   Score   Error  Units
BenchmarkLongDecimalWithOverflowAndLongStateSerializer.deserialize                  0               0         LongDecimalWithOverflowAndLongStateSerializer  avgt   20   4.787 ± 0.102  us/op
BenchmarkLongDecimalWithOverflowAndLongStateSerializer.deserialize                  0               0  LongDecimalWithOverflowAndLongStateSerializerRowType  avgt   20  12.861 ± 0.300  us/op
BenchmarkLongDecimalWithOverflowAndLongStateSerializer.serialize                    0               0         LongDecimalWithOverflowAndLongStateSerializer  avgt   20  12.929 ± 0.425  us/op
BenchmarkLongDecimalWithOverflowAndLongStateSerializer.serialize                    0               0  LongDecimalWithOverflowAndLongStateSerializerRowType  avgt   20  30.511 ± 1.516  us/op
```

Reply:
Those all seem like things we should improve in the RowType implementation. It would help in other places, too.

> minimal, dedicated block type.

Not just a block type, but a SQL type that's visible in the language, that needs to be maintained whenever changes we make changes to types and operators, etc.

BTW, I believe it's now possible to represent aggregation states as a union of multiple independent types that can get serialized separately without having to wrap with a ROW. cc @dain

Reply:
> Those all seem like things we should improve in the RowType implementation. It would help in other places, too.

How would you address 2) and 3)? It's not only optimization, but its a design question.

> Not just a block type, but a SQL type that's visible in the language, that needs to be maintained whenever changes we make changes to types and operators, etc.

I don't think it's visible in language (just like other internal types, e.g. `JoniRegexpType` or `Re2JRegexpType`). I don't think we even show intermediate aggregation type in explains.

> BTW, I believe it's now possible to represent aggregation states as a union of multiple independent types that can get serialized separately without having to wrap with a ROW. cc @dain

Is there an example of that? Would that be less complicated really?

Reply:
Take a look at RealAverageAggregation, which uses a BIGINT and a DOUBLE as intermediate types (https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/operator/aggregation/RealAverageAggregation.java#L71) and operates on two state objects: LongState and DoubleState (https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/operator/aggregation/RealAverageAggregation.java#L116)

There's no RowType, custom type of VARBINARY involved to serialized the intermediates. Each of the state objects gets serialized independently using the serializers for LongState and DoubleState.

Reply:
Actually, I think it moves problem level higher, since `RowType` will be used to serialized partial results anyway, see https://github.com/trinodb/trino/blob/992c7083d9fe5e49f2d161ab64049ef522bfecc2/core/trino-main/src/main/java/io/trino/sql/planner/LocalExecutionPlanner.java#L3613

Reply:
Another reason for not using custom types is that, eventually, we want to be able to push partial aggregations into connectors. The recent changes to how aggregations are declared, and the requirement for them to have a well-understood intermediate type, allows the connector to reason about the input/combine/output functions as regular functions and describe them to connectors for the purpose of pushdown. If we add hidden, engine-specific types for the intermediates, we won't be able to push partials into connectors.

Reply:
@lukasz-stec we should follow example from https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/operator/aggregation/RealAverageAggregation.java#L71 and optimize `RowType`

Reply:
I've crated issue for improving row type: https://github.com/trinodb/trino/issues/10354.
This PR should be only about improving decimal state and serialization